drivers: i2s: fix failure to restart I2S after a stop on Intel S1000

sathishkuttan opened the pull request:
This PR updates Intel S1000 I2S driver to fix the failure in restarting I2S after it has been stopped.

Commit:
drivers: i2s_cavs: add transmit underrun status

Add the definition for transmit underrun indicator bit in
I2S status register.

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi commented on soc/xtensa/intel_s1000/soc.h:
```diff
@@ -28,6 +28,17 @@
+#define CAVS_L2_AGG_INT_LEVEL3			DT_CAVS_ICTL_1_IRQ
+#define CAVS_L2_AGG_INT_LEVEL4			DT_CAVS_ICTL_2_IRQ
+#define CAVS_L2_AGG_INT_LEVEL5			DT_CAVS_ICTL_3_IRQ
+
```
It would be good if the changes to this file until the above point has its own commit. Then, in the subsequent commits, each of the interrupts should be defined in terms of  SOC_AGGREGATE_IRQ (like the way I2S interrupt is handled below).

sathishkuttan replied:
> It would be good if the changes to this file until the above point has its own commit.

I can do that.



> each of the interrupts should be defined in terms of SOC_AGGREGATE_IRQ (like the way I2S interrupt is handled below

I don't follow this comment. Are you asking to use the macro for defining the IRQ for other blocks as well -- the ones that you previously hard coded?

All of the IRQ definitions should go away from here and be moved to DTS, like what @galak has been doing, for example, in #12847 except that it should be done for all 3 values and them assembled using the macro.
It's for another PR.

rgundi replied:
> > It would be good if the changes to this file until the above point has its own commit.
> 
> I can do that.
> 
> > each of the interrupts should be defined in terms of SOC_AGGREGATE_IRQ (like the way I2S interrupt is handled below
> 
> I don't follow this comment. Are you asking to use the macro for defining the IRQ for other blocks as well -- the ones that you previously hard coded?
> 
Yes. That is what i meant.

> All of the IRQ definitions should go away from here and be moved to DTS, like what @galak has been doing, for example, in #12847 except that it should be done for all 3 values and them assembled using the macro.
> It's for another PR.
Yes. We wanted this to be done in DTS too but that involved changing the scripts which generates the macros and we were not so sure about it. Moreover, it was only s1000 which was using a multilevel interrupt structure at that time. It may now be worthwhile to do that as there are a couple more SoCs who are using the multilevel interrupt feature.

sathishkuttan replied:
I already split the macro definition and the usage for I2S.
Maybe you can take care of the move to DTS.
See https://github.com/zephyrproject-rtos/zephyr/pull/12847#issuecomment-459482382

Commit:
soc: intel_s1000: Macro to encode multi-level IRQ

Add Macro to encode multi-level aggregation of interrupts
into an encoded IRQ value

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

Commit:
soc: intel_s1000: Encoded IRQ definitions for I2S

Use IRQ encoding macro to define the IRQ identifiers for I2S

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi reviewed (commented):
Not sure i understand all the changes done in Commit 4. May be splitting it up into multiple commits help?

sathishkuttan commented:
> Not sure i understand all the changes done in Commit 4. May be splitting it up into multiple commits help?

If the commit message doesn't help, then I doubt splitting into multiple commits is going to help.

rgundi commented:
Well.. the standard practice is to split it into multiple commits so that reviewing becomes easier as the reviewer will have less changes to see at a time. I see too many changes to one file here and that's scaring me off. Anyway, it was just a suggestion.

sathishkuttan commented:
> Well.. the standard practice is to split it into multiple commits so that reviewing becomes easier as the reviewer will have less changes to see at a time. I see too many changes to one file here and that's scaring me off. Anyway, it was just a suggestion.

I'll try to split into 2 commits. No more granular than that.

Commit:
drivers: i2s_cavs: reduce buffer queue length

Reduce the length of internal buffer queue from 4 down to 2.

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

Commit:
drivers: i2s_cavs: macros for object instantiation

Add macros to define and instantiate driver objects
for multiple instances

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

Commit:
drivers: i2s_cavs: fix for I2S start after stop

Fixed an issue I2S wouldn't resume when started after a stop
Added code to empty the TX/RX FIFOs upon stop
TX stop is achieved by letting FIFO underrun, then changing state in ISR
RX FIFO is read until empty when RX is stopped

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi reviewed (approved):
LGTM
